### PR TITLE
make chdir() return true/false as its documentation claims

### LINKED ISSUE #22365: chdir doesn't return a boolean, but manual says it returns true/false
**Description**

According to `perldoc -f chdir`:

> It returns true on success, false otherwise.

However, `chdir` actually returns (integer) 1 and 0. This is mostly observable in the failure case, which (according to the documentation) should return a value that is `eq ""`, but isn't.

**Steps to Reproduce**
```
$ perl -wE 'say chdir "/invalid*"'
0
$ perl -Mexperimental=builtin -wE 'say builtin::is_bool chdir "/invalid*"'

$
```

**Expected behavior**
```
$ perl -wE 'say chdir "/invalid*"'

$ perl -Mexperimental=builtin -wE 'say builtin::is_bool chdir "/invalid*"'
1
$
```
I.e. a false value should stringify to the empty string and be `is_bool`.

Alternatively, the documentation should be changed to match the existing behavior.

(This is with v5.40.0.)

### PR DESCRIPTION
**Pull request by mauke**
Previously it would return integer 1 and 0, not booleans.
    
Fixes #22365.

### CODE REVIEW

File: lib/B/Op_private.pm [hunk 1]
```diff
@@ -150,7 +150,7 @@ $bits{$_}{6} = 'OPpPAD_STATE' for qw(emptyavhv lvavref lvref padav padhv padsv p
...
-$bits{$_}{4} = 'OPpTARGET_MY' for qw(abs add atan2 ceil chdir chmod chomp chown chr chroot concat cos crypt divide emptyavhv exec exp flock floor getpgrp getppid getpriority hex i_add i_divide i_modulo i_multiply i_negate i_subtract index int kill left_shift length link log mkdir modulo multiconcat multiply nbit_and nbit_or nbit_xor ncomplement negate oct ord pow push rand refaddr reftype rename right_shift rindex rmdir schomp scomplement setpgrp setpriority sin sleep sqrt srand stringify subtract symlink system time undef unlink unshift utime wait waitpid);
+$bits{$_}{4} = 'OPpTARGET_MY' for qw(abs add atan2 ceil chmod chomp chown chr chroot concat cos crypt divide emptyavhv exec exp flock floor getpgrp getppid getpriority hex i_add i_divide i_modulo i_multiply i_negate i_subtract index int kill left_shift length link log mkdir modulo multiconcat multiply nbit_and nbit_or nbit_xor ncomplement negate oct ord pow push rand refaddr reftype rename right_shift rindex rmdir schomp scomplement setpgrp setpriority sin sleep sqrt srand stringify subtract symlink system time undef unlink unshift utime wait waitpid);
```
Comment (jkeenan) [inline lib/B/Op_private.pm:153]: @mauke, in this p.r. you remove `chdir` from two lists (here and at line 909 below) which are the object of key-value pairs in lib/B/Op_private.pm.  Could you give us a sense of what these lists are for in the first place, and why removing `chdir` from them is the right thing to do?

File: opcode.h [hunk 2]
```diff
@@ -3028,7 +3028,7 @@ EXTCONST U16  PL_op_private_bitdefs[] = {
...
     0x06d4, 0x0770, 0x0003, /* ref, blessed */
-    0x018f, /* bless, glob, sprintf, formline, unpack, pack, join, anonlist, anonhash, splice, warn, die, reset, exit, close, pipe_op, fileno, umask, binmode, tie, dbmopen, sselect, select, getc, read, enterwrite, sysopen, sysseek, sysread, syswrite, eof, tell, seek, truncate, fcntl, ioctl, send, recv, socket, sockpair, bind, connect, listen, accept, shutdown, gsockopt, ssockopt, open_dir, seekdir, gmtime, shmget, shmctl, shmread, shmwrite, msgget, msgctl, msgsnd, msgrcv, semop, semget, semctl, ghbyaddr, gnbyaddr, gpbynumber, gsbyname, gsbyport, syscall */
+    0x018f, /* bless, glob, sprintf, formline, unpack, pack, join, anonlist, anonhash, splice, warn, die, reset, exit, close, pipe_op, fileno, umask, binmode, tie, dbmopen, sselect, select, getc, read, enterwrite, sysopen, sysseek, sysread, syswrite, eof, tell, seek, truncate, fcntl, ioctl, send, recv, socket, sockpair, bind, connect, listen, accept, shutdown, gsockopt, ssockopt, chdir, open_dir, seekdir, gmtime, shmget, shmctl, shmread, shmwrite, msgget, msgctl, msgsnd, msgrcv, semop, semget, semctl, ghbyaddr, gnbyaddr, gpbynumber, gsbyname, gsbyport, syscall */
```
Comment (jkeenan) [inline opcode.h:3031]: Similarly, here in opcode.h you are adding `chdir` to a list where it was not previously present, while at line 3044 below you are removing `chdir` from a list.  Can you provide the layperson what the functions of these two lists are?

File: lib/B/Op_private.pm (hunk 1 above)
Comment (mauke) [reply lib/B/Op_private.pm:153]: Dunno, lol. This file is autogenerated and (presumably) contains metadata for various ops.

What I did was to remove the `T` flag from `chdir` in `regen/opcodes`. `chdir` used to have a "target" (which seems to be a pre-allocated return value slot) for storing the integer it was going to return. Now that `chdir` only ever returns canonical booleans (`PL_sv_yes` and `PL_sv_no`), it doesn't need a target anymore. (That's why `dTARGET` is gone from the op code.)

That means we don't need to pre-allocate a target slot for `chdir` anymore, which in turn is controlled by `PL_opargs[type] & OA_TARGET` (see e.g. `Perl_newOP` in `op.c`). The `PL_opargs` array is initialized in `opcode.h`, which in turn is generated from the data in `regen/opcodes`.

Presumably `B` needs similar metadata to traverse and inspect ops, so it needs to know which ones can have targets.

File: opcode.h (hunk 2 above)
Comment (mauke) [reply opcode.h:3031]: See my comment above.

These are just comments. Presumably they are meant to indicate to the reader which sets of ops share the same set of allowed flags that they can have.

File: ext/POSIX/t/wrappers.t [hunk 3]
```diff
@@ -164,7 +164,7 @@ SKIP: {
 
 is(-e NOT_HERE, undef, NOT_HERE . ' does not exist');
 
-foreach ([undef, 0, 'chdir', NOT_HERE],
+foreach ([undef, $^V ge v5.41.1 ? !1 : 0, 'chdir', NOT_HERE],
```
Comment (Leont) [inline ext/POSIX/t/wrappers.t]: POSIX isn't dual-life so this shouldn't be necessary

File: pp_sys.c [hunk 4]
```diff
@@ -3972,7 +3972,7 @@ PP_wrapped(pp_chdir, MAXARG, 0)
 #endif
     }
     else 
-        PUSHi( PerlDir_chdir(tmps) >= 0 );
+        PUSHs(boolSV( PerlDir_chdir(tmps) >= 0 ));
```
Comment (Leont) [inline pp_sys.c:3975]: This feels like we should add a `PUSHb` macro.

File: ext/POSIX/t/wrappers.t (hunk 3 above)
Comment (mauke) [reply ext/POSIX/t/wrappers.t]: Good point. I've removed the version check.

File: pp_sys.c (hunk 4 above)
Comment (mauke) [reply pp_sys.c:3975]: Perhaps, but aren't the `PUSH*` macros going to go away once we have a fully refcounted stack?

(Anyway, this is definitely out-of-scope for this PR.)

**Comment by tonycoz** [comment]
This looks reasonable to me but this looks like it will conflict with the other two pp_chdir PRs.

**Comment by mauke** [comment]
I've rebased this PR on top of the `&CORE::chdir` change. The `pp_chdir` code itself had no conflicts, but one of the tests had to be adapted.

The remaining PR (#22368) will have code conflicts, but they're fairly straightforward to resolve (even if git doesn't know how).

**Comment by iabyn** [comment]
On Mon, Jul 01, 2024 at 08:52:28AM -0700, mauke wrote:
> What I did was to remove the `T` flag from `chdir` in `regen/opcodes`.
> `chdir` used to have a "target" (which seems to be a pre-allocated
> return value slot) for storing the integer it was going to return. Now
> that `chdir` only ever returns canonical booleans (`PL_sv_yes` and
> `PL_sv_no`), it doesn't need a target anymore. (That's why `dTARGET` is
> gone from the op code.)

The T flag is a bit more than that. It indicates that an optimisation can
be used whereby an assignment to a lexical var is optimised away. So for
example, with

    my $x;
    $x = chdir(...),

*without* the optimisation, chdir would set the PADTMP allocated for it's
benefit to 0 or 1, then push it on the stack. Then $x would be pushed on
the stack by the OP_PADSV, then the OP_SASSIGN  would pop the two SVs off
the stack, and assign the value of one to the other, thus setting $x to 0
or 1.

*With* the optimisation, the OP_PADSV and OP_SASSIGN are optimised away,
and the OP_CHDIR is modified so that the index in the pad of its "PADTMP"
is actually the index of the $x lex var. Thus when pp_chdir() sets its
"PADTMP" to 0 or 1, its actually setting $x directly.

This is very useful optimisation, commonly used in things like arithmetic
ops, so $x = $a + $b for example optimises away the assignment.

The optimisation isn't relevant if a lexical isn't being assigned to, for
example the common idiom

    chdir ....
        or die ....;

Whether lexical assignments are common enough, or whether chdir() has
enough OS overhead that such an optimisation isn't needed, are open
questions.

-- 
A walk of a thousand miles begins with a single step...
then continues for another 1,999,999 or so.

**Comment by mauke** [comment]
> The T flag is a bit more than that. It indicates that an optimisation can be used whereby an assignment to a lexical var is optimised away.

Ah, interesting. I didn't know that.

As for performance considerations, I don't have any hard data. None of my code uses `chdir` anywhere near an inner loop.

But I did run an approximate search for `chdir` on the perl source tree and found about 1600 results. Some of those are false positives, though: occurrences of the word `chdir` in strings, comments, documentation, etc. Still, I would estimate we have on the order of 1000 `chdir` calls in Perl code.

Of those, exactly one involves an assignment to a lexical variable:
```perl
        $ok = chdir($updir_loc) unless ($p_dir eq $File::Find::current_dir);
```
in `ext/File-Find/lib/File/Find.pm`, line 567.

(That's from before applying this PR, which adds a few `$r = chdir(...);` statements in `t/op/chdir.t`.)

File::Find is an extreme outlier: It contains nine distinct calls to `chdir`, but even so, eight of them are in boolean context (in `or`, `if`, or `unless`). Similarly, the other `chdir` calls in the perl source are either in boolean or void context.

My conclusion is that this optimization practically never applies to real-world code, and where it does, it is most likely dwarfed by the context switch to kernel mode and touching the file system.

**Review by tonycoz — APPROVED** [review]

**Pull request merged**